File: backend/app/schemas/experiment.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ExperimentCreate(BaseModel):
    name: str
    stream_id: str
    protocol: str
    network_profile: str = "normal"
    encoder_params: str | None = None
    avg_latency_ms: float | None = None
    max_latency_ms: float | None = None
    p50_latency_ms: float | None = None
    p95_latency_ms: float | None = None
    p99_latency_ms: float | None = None
    sample_count: int = 0
    measurement_method: str | None = None
    bitrate_kbps: float | None = None
    fps: float | None = None
    stutter_count: int = 0


class ExperimentRead(ExperimentCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime

File: frontend/src/api/experiment.ts
import { apiGet, apiPost } from "./client";

export interface Experiment {
  id: number;
  name: string;
  stream_id: string;
  protocol: string;
  network_profile: string;
  encoder_params?: string | null;
  avg_latency_ms?: number | null;
  max_latency_ms?: number | null;
  p50_latency_ms?: number | null;
  p95_latency_ms?: number | null;
  p99_latency_ms?: number | null;
  sample_count: number;
  measurement_method?: string | null;
  bitrate_kbps?: number | null;
  fps?: number | null;
  stutter_count: number;
  created_at: string;
}

export function listExperiments() {
  return apiGet<Experiment[]>("/experiments");
}

export function createLatencyExperiment(payload: Omit<Experiment, "id" | "created_at">) {
  return apiPost<Experiment>("/experiments/latency", payload);
}
